File: trunk/jsVortex/socket-connector/JavaSocketConnector.java
/**
 ** Copyright (C) 2009 Advanced Software Production Line, S.L.
 ** See license.txt or http://www.aspl.es/vortex
 **/

import java.applet.*;
import javax.swing.*;
import netscape.javascript.*;
import java.net.*;
import java.io.*;

public class JavaSocketConnector extends JApplet {

	/* A reference to the current browser (tab) opening the
	 * component */
	JSObject      browser      = null; /* browser */
	BlockingQueue commandQueue = null; /* command queue */

	/* list of callers that are inside the applet */
	Callers       callers      = null;
	boolean       running;
	

	/**
	 * Public initialization. Get a reference to the browser
	 * initializing the applet.
	 */
	public void init() {
		/* reference to the browser */
		browser = JSObject.getWindow (this);

		/* initialize callers */
		callers = new Callers ();

		return;
	}

	/** 
	 * @brief Stop the applet.
	 */
	public void stop () {
		running = false;
	}
	public void destroy () {
		stop ();
	}

	/** 
	 * @brief Main loop that waits for commands to be implemented
	 * under this thread because it has permission.
	 */
	public void start () {
		/* Notify the browser that the component was
		 * loaded. */
		JSObject member = (JSObject) browser.getMember ("JavaSocketConnector");
		member.setMember ("isReady", true);

		/* create commandQueue */
		commandQueue = new BlockingQueue ();

		running = true;
		Command cmd = null;

		/* set lowest priority */
		Thread.currentThread ().setPriority (Thread.MIN_PRIORITY);

		while (running){
			try {
				/* notify caller inside */
				synchronized (callers) {
					/* check if the caller list if
					 * empty (no javaScript thread
					 * is inside the applet) */
					if (callers.count > 0) 
						callers.wait ();
				} /* end if */
				/* give a try to other threads */
				Thread.yield ();

				/* Wait for the next operation requested */
				cmd = (Command) commandQueue.pop ();
			} catch (Exception ex) {
				/* do some log error here */
				continue;
			} /* end try */

			/* call to complete command */
			cmd.doOperation (browser, this);
		}  /* end while */
		return;
	}

	/** 
	 * @brief Socket connect to the host and port provided. Once
	 * the connection is created, it is notified on the provided
	 * handler.
	 * 
	 * @param host The host to connect to.
	 * @param port The port to connect to.
	 */
	public boolean connect (String host, int port, JSObject caller) {

		/* notify caller inside */
		synchronized (callers) {
			callers.count++;
		}

		/* create the socket command */
		SocketCommand cmd = new SocketCommand ();
		cmd.host   = host;
		cmd.port   = port;
		cmd.caller = caller;

		/* queue the command */
		commandQueue.push (cmd);

		/* notify caller inside */
		synchronized (callers) {
			callers.count--;
			callers.notify ();
		}
		
		return true;
	}

	/** 
	 * @brief Allows to send the provided string over the provided
	 * out stream (associated to a particular socket).
	 *
	 * @param content The content to be sent.
	 * @param length The amount of data to be written.
	 * @param out The output stream object to write on.
	 */
	public boolean send (String content, int length, PrintWriter output, JSObject caller){

		/* notify caller inside */
		synchronized (callers) {
			callers.count++;
		}

		/* queue a send operation */
		SendCommand sendCmd = new SendCommand ();
		sendCmd.content = content;
		sendCmd.length  = length;
		sendCmd.output  = output;
		sendCmd.caller  = caller;

		/* queue command */
		commandQueue.push (sendCmd);

		/* notify caller inside */
		synchronized (callers) {
			callers.count--;
			callers.notify ();
		}

		return true;
	}

	/** 
	 * @brief Activates TLS support on the provided socket object
	 * (caller reference).
	 */
	public boolean enableTLS (JSObject caller) {
		/* notify caller inside */
		synchronized (callers) {
			callers.count++;
		}

		/* call to create command */
		EnableTLSCommand cmd = new EnableTLSCommand ();
		cmd.caller           = caller;

		commandQueue.push (cmd);

		/* notify caller inside */
		synchronized (callers) {
			callers.count--;
			callers.notify ();
		}

		return true;
	}

	/** 
	 * @brief Closes the socket by closing internal socket, output
	 * stream and input stream. The method also changes the
	 * readyState of the socket and fires the onclose event.
	 *
	 * @param caller The caller and at the same time the socket.
	 */
	public void close (JSObject caller) {

		/* notify caller inside */
		synchronized (callers) {
			callers.count++;
		}

		PrintWriter    out      = (PrintWriter) caller.getMember ("_jsc_out");
		caller.removeMember ("_jsc_out");

		Socket         socket   = (Socket) caller.getMember ("_jsc_socket");
		caller.removeMember ("_jsc_socket");

		SocketListener listener = (SocketListener) caller.getMember ("_jsc_listener");
		caller.removeMember ("_jsc_listener");

		/* close all items */
		listener.close ();
		try {out.close ();} catch (Exception ex) {}

		/* now change ready state */
		caller.setMember ("readyState", 2);

		/* fire onclose event */
		notify (caller, "onclose", null);

		/* notify caller inside */
		synchronized (callers) {
			callers.count--;
			callers.notify ();
		}
		
		return;
	}

	public void notify (JSObject caller, String handler, Object arg) {
		/* LogHandling.info (caller, "Doing handler notification for: " + _handler);  */

		/* notify caller inside */
		synchronized (callers) {
			callers.count++;
		}

		InvokeCommand invokeCmd = new InvokeCommand ();
		invokeCmd.handler = handler;
		invokeCmd.caller  = caller;
		invokeCmd.arg     = arg;

		/* push command */
		commandQueue.push (invokeCmd);

		/* notify caller inside */
		synchronized (callers) {
			callers.count--;
			callers.notify ();
		}
	}

} /* end JavaSocketConnector */

File: trunk/jsVortex/socket-connector/JavaSocketConnector.js
/**
 ** Copyright (C) 2009 Advanced Software Production Line, S.L.
 ** See license.txt or http://www.aspl.es/vortex
 **/

/**
 * @internal The following is defined to avoid crashing js code in the
 * case console module is not defined.
 */
if (typeof console == "undefined") {
    this.console = {
	log:   function () {},
	error: function () {},
	warn:  function () {}
    };
}

/**
 * @brief Socket constructor. The constructor returns a Socket
 * object. The connection will be notified at the onopen method or
 */
function JavaSocketConnector (params) {
    /**
     * @brief Reference to the host the socket connects to.
     */
    this.host = params.host;
    /**
     * @brief Reference to the port the socket connects to.
     */
    this.port = params.port;

    /**
     * @brief Connection status. By default it is set to CONNECTING =
     * 0. The list of readyState are:
     *  - CONNECTING = 0
     *  - OPEN       = 1
     *  - CLOSED     = 2
     */
    this.readyState = 0;

    /* do a socket connection */
    document.getElementById('JavaSocketConnector').connect (params.host, params.port, this);
}

/**
 * @brief Global variable used to signal that the applet was loaded
 * and started.
 */
JavaSocketConnector.isReady = false;

/**
 * @brief Allows to send content over the provided socket object.
 *
 * @param content The content to be sent.
 * @param length The amount of data to be sent from the content.
 *
 * @return true in the case the send operation was initiated,
 * otherwise false is returned.
 */
JavaSocketConnector.prototype.send = function (content, length) {
    /* check socket readyState */
    if (this.readyState != 1) {
	this.onlog ("error", "Unable to send content, socket readyState is: " + readyState);
	return false;
    }

    /* now send content */
    return document.getElementById('JavaSocketConnector').send (content, length, this._jsc_out, this);
};

/**
 * @brief Function used to enable TLS protection on the provided socket.
 * See also certTrustPolicy which allows configuring what to do in the case of certificate error.
 */
JavaSocketConnector.prototype.enableTLS = function () {
    /* check socket readyState */
    if (this.readyState != 1) {
	this.onlog ("error", "Unable to enable TLS, socket readyState is: " + readyState);
	return false;
    }

    /* now send content */
    return document.getElementById('JavaSocketConnector').enableTLS (this);
};

/**
 * @brief Certificate trust policy for TLS activation. This member
 * allows to configure what to do in the case server certificate
 * validation fails. Allowed values are:
 * 1 : Only accept valid server certificates.
 * 2 : Ask the user (caller) to accept or not a certificate when an error is found. The caller must have defined oncerterror method. See it for details.
 * 3 : In the case of certificate error, accept it.
 */
JavaSocketConnector.prototype.certTrustPolicy = 1;

JavaSocketConnector.prototype.close = function () {
    if (this.readyState == 2) {
	this.onlog ("warn", "Connection already closed");
	return;
    }

    /* call to close */
    document.getElementById('JavaSocketConnector').close (this);
    return;
};

/**
 * @brief Default method to get connection open notifications.
 */
JavaSocketConnector.prototype.onopen = function () {
    if (this.readyState == 1) {
	console.log ("USING DEFAULT onopen: Socket connected to: " + this.host + ", port: " + this.port);
	return;
    }
    console.error ("USING DEFAULT onopen: Failed to connect!");
};

/**
 * @brief This is the handler that will receive all content received
 * on the socket.
 *
 * @param content The content received over the socket.
 */
JavaSocketConnector.prototype.onmessage = function (content) {
    console.log ("USING DEFAULT onmessage: Content received: " + content);
};

/**
 * @brief This is the handler that will receive on close notifications.
 */
JavaSocketConnector.prototype.onclose = function () {
    console.log ("USING DEFAULT onclose: Close notification");
};

/**
 * @brief This is the handler that will be called once the TLS handshake have finished.
 */
JavaSocketConnector.prototype.ontls = function (status) {
    console.log ("USING DEFAULT ontls: TLS status was: " + status);
};

/**
 * @brief Handler called in the case TLS is activated and a server
 * certificate error is found.
 */
JavaSocketConnector.prototype.oncerterror = function (subjet, issuer, cert) {
    console.log ("USING DEFAULT oncerterror: validating subject: " + subject);
    /* this default implementation never accept a certificate */
    return false;
};

/**
 * @brief Default onlog handler used by the JavaSocketConnector applet
 * to relay all logs created during its function.
 */
JavaSocketConnector.prototype.onlog = function (type, message) {

    if (type == "info") {
	console.log (message);
	return;
    } else if (type == "error") {
	console.error (message);
	return;
    } else if (type == "warn") {
	console.warn (message);
	return;
    }

    console.error ("UNHANDLED TYPE: " + type + ": " + message);
    return;
};

/**
 * @internal Handler used by the java accept to marshall the call.
 */
JavaSocketConnector.prototype._marshallCall = function (context, call, arg) {
    call.apply (context, [arg]);
};


